Bottom: d6abc36565981fa2846a35ed6055d9133f877bac
Top:    a1fd62ff5f6841cef7d4d0fbc516e47ec95e7f7f
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-26 14:52:37 +0000

Refresh of try-consolidating-geometry-in

---

diff --git a/src/hid/common/object3d.c b/src/hid/common/object3d.c
index 494b9c7..149dcfe 100644
--- a/src/hid/common/object3d.c
+++ b/src/hid/common/object3d.c
@@ -110,19 +110,19 @@ object3d_set_appearance (object3d *object, appearance *appear)
 void
 object3d_add_edge (object3d *object, edge_ref edge)
 {
-  object->edges = g_list_append (object->edges, (void *)edge);
+  object->edges = g_list_prepend (object->edges, (void *)edge);
 }
 
 void
 object3d_add_vertex (object3d *object, vertex3d *vertex)
 {
-  object->vertices = g_list_append (object->vertices, vertex);
+  object->vertices = g_list_prepend (object->vertices, vertex);
 }
 
 void
 object3d_add_face (object3d *object, face3d *face)
 {
-  object->faces = g_list_append (object->faces, face);
+  object->faces = g_list_prepend (object->faces, face);
 }
 
 
@@ -680,7 +680,7 @@ object3d_from_contours (POLYAREA *contours,
       splice (cylinder_edges[1], SYM(cylinder_edges[1]));
     }
 
-    objects = g_list_append (objects, object);
+    objects = g_list_prepend (objects, object);
 
     link = malloc (sizeof (polygon_3d_link));
     pa->user_data = link;
@@ -1359,7 +1359,7 @@ old_object3d_from_board_outline (void)
       splice (cylinder_edges[1], SYM(cylinder_edges[1]));
     }
 
-    board_objects = g_list_append (board_objects, board_object);
+    board_objects = g_list_prepend (board_objects, board_object);
 
   } while (pa = pa->f, pa != board_outline);
 
diff --git a/src/hid/step/step_writer.c b/src/hid/step/step_writer.c
index 178d8f3..0ed1db3 100644
--- a/src/hid/step/step_writer.c
+++ b/src/hid/step/step_writer.c
@@ -69,12 +69,18 @@ step_file
   file->f = f;
   file->next_id = 1;
 
+  file->cartesian_point_hash = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  file->direction_hash =       g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  file->vector_hash =          g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
   return file;
 }
 
 void
 destroy_step_output_file (step_file *file)
 {
+  g_hash_table_destroy (file->cartesian_point_hash);
+  g_hash_table_destroy (file->direction_hash);
   g_free (file);
 }
 
@@ -189,16 +195,42 @@ step_product_definition_shape (step_file *file, char *name, char *description, s
 step_id
 step_cartesian_point (step_file *file, char *name, double x, double y, double z)
 {
-  fprintf (file->f, "#%i=CARTESIAN_POINT('%s',(%f,%f,%f));\n",
-                    file->next_id, name, x, y, z);
+  char *line;
+  step_id id;
+
+  line = g_strdup_printf ("'%s',(%f,%f,%f)", name, x, y, z);
+
+  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (file->cartesian_point_hash, line))) != 0)
+    {
+      g_free (line);
+      return id;
+    }
+
+  g_hash_table_insert (file->cartesian_point_hash, line, GINT_TO_POINTER (file->next_id));
+  fprintf (file->f, "%i=CARTESIAN_POINT(%s);\n", file->next_id, line);
+  g_free (line);
+
   return file->next_id++;
 }
 
 step_id
 step_direction (step_file *file, char *name, double x, double y, double z)
 {
-  fprintf (file->f, "#%i=DIRECTION('%s',(%f,%f,%f));\n",
-                    file->next_id, name, x, y, z);
+  char *line;
+  step_id id;
+
+  line = g_strdup_printf ("'%s',(%f,%f,%f)", name, x, y, z);
+
+  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (file->direction_hash, line))) != 0)
+    {
+      g_free (line);
+      return id;
+    }
+
+  g_hash_table_insert (file->direction_hash, line, GINT_TO_POINTER (file->next_id));
+  fprintf (file->f, "%i=DIRECTION(%s);\n", file->next_id, line);
+  g_free (line);
+
   return file->next_id++;
 }
 
@@ -237,8 +269,21 @@ step_circle (step_file *file, char *name, step_id position, double radius)
 step_id
 step_vector (step_file *file, char *name, step_id orientation, double magnitude)
 {
-  fprintf (file->f, "#%i=VECTOR('%s',#%i,%f);\n",
-                    file->next_id, name, orientation, magnitude);
+  char *line;
+  step_id id;
+
+  line = g_strdup_printf ("'%s',#%i,%f", name, orientation, magnitude);
+
+  if ((id = GPOINTER_TO_INT (g_hash_table_lookup (file->vector_hash, line))) != 0)
+    {
+      g_free (line);
+      return id;
+    }
+
+  g_hash_table_insert (file->vector_hash, line, GINT_TO_POINTER (file->next_id));
+  fprintf (file->f, "%i=VECTOR(%s);\n", file->next_id, line);
+  g_free (line);
+
   return file->next_id++;
 }
 
diff --git a/src/hid/step/step_writer.h b/src/hid/step/step_writer.h
index c86aa1e..1702b77 100644
--- a/src/hid/step/step_writer.h
+++ b/src/hid/step/step_writer.h
@@ -5,6 +5,9 @@ typedef GList* step_id_list;
 typedef struct {
   FILE *f;
   step_id next_id;
+  GHashTable *cartesian_point_hash;
+  GHashTable *direction_hash;
+  GHashTable *vector_hash;
 
 } step_file;
